src/client.ts

```diff
@@ -0,0 +1,217 @@
+};
+
+export type ConduitClientConfig<T> = {
+  orgId: number;
+  streamUrl: string;
+  baseConduitUrl: string;
+  startStreamData?: Record<string, unknown>;
+  onMessage?: (message: T) => void;
+  onOpen?: () => void;
+  onClose?: () => void;
+  onError?: (err: Error) => void;
+};
+
+export class ConduitClient<T> {
+  private config: ConduitClientConfig<T>;
+  private eventSource: EventSource | null = null;
+  private connecting = false;
+
+  private lastEventId: string | undefined;
+  private lastSeq: number | undefined;
+  private seenIds = new Set<string>();
+
+  private tokenExpiresAt: number | undefined;
+  private currentChannelId: string | undefined;
+
+  constructor(config: ConduitClientConfig<T>) {
+    this.config = config;
+    this.handleStream = this.handleStream.bind(this);
+    this.handleControl = this.handleControl.bind(this);
+  }
+
+  private async startStream(): Promise<StartStreamResponse> {
+    const response = await fetch(this.config.streamUrl, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        org_id: this.config.orgId,
+        ...this.config.startStreamData,
+      }),
+    });
+
+    if (!response.ok) {
+      throw new Error(`Failed to start stream ${response.statusText}`);
+    }
+
+    return response.json();
+  }
+
+  private buildUrl(token: string, channelId: string): string {
+    const queryParams = new URLSearchParams({
+      token,
+      channel_id: channelId,
+    });
+    if (this.lastEventId) queryParams.set('last_event_id', this.lastEventId);
+    return `${this.config.baseConduitUrl.replace(/\/$/, '')}/events/${this.config.orgId}?${queryParams.toString()}`;
+  }
+
+  private attach(url: string) {
+    if (this.tokenExpiresAt && Date.now() >= this.tokenExpiresAt) {
+      return;
+    }
+
+    this.eventSource = new EventSource(url);
+
+    this.eventSource.onopen = () => {
+      this.config.onOpen?.();
+    };
+
+    if (this.tokenExpiresAt && Date.now() >= this.tokenExpiresAt) {
+      const e: Error & { code?: string } = new Error('Token expired at reconnect');
+      e.code = 'TOKEN_EXPIRED';
+      this.config.onError?.(e);
+      this.disconnect();
+      return;
+    }
+
+    this.eventSource.onerror = () => {
+      if (this.eventSource == null) return;
+      this.config.onError?.(new Error('SSE connection error'));
+    };
+
+    this.eventSource.addEventListener('stream', this.handleStream);
+    this.eventSource.addEventListener('control', this.handleControl);
+  }
+
+  async connect(): Promise<void> {
+    if (this.connecting || this.eventSource) return;
+    this.connecting = true;
+    try {
+      const { token, channel_id, expires_in } = await this.startStream();
+
+      if (channel_id !== this.currentChannelId) {
+        this.currentChannelId = channel_id;
+        this.lastSeq = undefined;
+        this.seenIds.clear();
+      }
+
+      if (Number.isFinite(expires_in) && expires_in > 0) {
```

Comment:
Do we actually need this? The types specify that this will always be present and I don't think we need to worry about this being a huge number.